### PR TITLE
Don't use blocking connections on the main thread

### PR DESCRIPTION
**Pull request by jherico**
We're still seeing a lot of deadlocks in the application and crashes in QtScript / QtQml.

Some of this is potentially coming from the `Audio` scripting interface on the main thread calling blocking functions on the `AudioClient` (which runs on it's own thread).  

Some of it is also potentially from threading issues wherever scripts interact with QML in a non-thread safe way, such as in QML based overlays.  

This PR makes several changes.  

* Adds new helper functions and macros for making blocking method invocations.
  * These emit log warning if they occur from the main thread
  * Macro usage ensures that the log warning include the calling function in which the blocking was made
* Migrates _all_ use of `QMetaObject::invokeMethod` that uses `Qt::BlockingQueuedConnection` to use the new methods / macros
* Modified Audio scripting interface to not use blocking functions when communicating with the AudioClient.  
* Modifies Overlay instances to always live on the main thread.  All interaction with the Overlay scripting interface now funnels through the main thread.
* Modifies OvlerayWindows and OverlayWebWindows to live on the main thread.  
* Updates many scripting interfaces to use the self-invocation pattern when they are called from another thread
* Eliminates the unused OverlayPanel code (using the preprocessor so that it can be restored easily if there is some user script taking advantage of it)
* Eliminates support for the tabbed tool window, previously only used by the edit.js code

## Testing

* Verify startup and shutdown function without hanging or crashing
* Verify Overlay functionality still works.  (this [script](https://raw.githubusercontent.com/jherico/DreamingScripts/master/tests/overlays.js) creates a number of overlays)
* Verify audio settings dialog still functions properly (check boxes work, changing input and output devices works)
* Smoke test functionality
* Monitor log output for warnings containing the string `BlockingQueuedConnection` (this doesn't fail the PR, it just informative and should give us clues to additional sources of race conditions)

### CODE REVIEW

File: interface/src/scripting/AudioDevices.cpp [hunk 1]
```diff
@@ -71,22 +73,14 @@ bool AudioDeviceList::setData(const QModelIndex& index, const QVariant& value, i
...
     // skip if already selected
     if (!device.selected) {
         auto client = DependencyManager::get<AudioClient>();
-        QMetaObject::invokeMethod(client.data(), "switchAudioDevice", Qt::BlockingQueuedConnection,
-            Q_RETURN_ARG(bool, success),
+        QMetaObject::invokeMethod(client.data(), "switchAudioDevice",
             Q_ARG(QAudio::Mode, _mode),
             Q_ARG(const QAudioDeviceInfo&, device.info));
-
-        if (success) {
```
Comment (jherico) [inline interface/src/scripting/AudioDevices.cpp]: rely the deviceChanged signal to modify our internal state.

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/1221/HighFidelity-Beta-PR10860-7073dc301c64a6871f4fabc45032f3aa60b65f39.dmg

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/1221/HighFidelity-Beta-PR10860-7073dc301c64a6871f4fabc45032f3aa60b65f39.exe

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/1222/HighFidelity-Beta-PR10860-3669618b6b8947ac1146a0ffbb69a7864829ee3a.dmg

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/1222/HighFidelity-Beta-PR10860-3669618b6b8947ac1146a0ffbb69a7864829ee3a.exe

**Comment by sethalves** [comment]
01:16:08 /home/jenkins/workspace/pr-build/label/ubuntu/libraries/audio/src/SoundCache.cpp:14:26: fatal error: QtCore/Qthread: No such file or directory
01:16:08 compilation terminated.

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/1223/HighFidelity-Beta-PR10860-d3b9958b270f9227406adca93d3f4fb013abbe4a.exe

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/1223/HighFidelity-Beta-PR10860-d3b9958b270f9227406adca93d3f4fb013abbe4a.dmg

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/1228/HighFidelity-Beta-PR10860-8372d73fec32336f3b6dc20f06909f930601f320.exe

**Comment by YuppyPlays** [comment]
Will test once built :)

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/1236/HighFidelity-Beta-PR10860-9c37b28ba0b96430ec7f14019da18363ed4ccda6.dmg

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/1236/HighFidelity-Beta-PR10860-9c37b28ba0b96430ec7f14019da18363ed4ccda6.exe

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/1240/HighFidelity-Beta-PR10860-6fc82bb351626d901aa060009dafc67d25c15683.dmg

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/1240/HighFidelity-Beta-PR10860-6fc82bb351626d901aa060009dafc67d25c15683.exe

File: interface/src/scripting/Audio.cpp [hunk 2]
```diff
@@ -93,19 +95,11 @@ void Audio::setInputVolume(float volume) {
 
     if (_inputVolume != volume) {
         auto client = DependencyManager::get<AudioClient>().data();
-        QMetaObject::invokeMethod(client, "setInputVolume", Qt::BlockingQueuedConnection, Q_ARG(float, volume));
```
Comment (jherico) [inline interface/src/scripting/Audio.cpp]: dont' rely on a blocking queued connection... instead rely on a new `inputVolumeChanged` signal from the AudioClient to update local state

File: interface/src/ui/overlays/Overlays.h [hunk 3]
```diff
@@ -316,13 +321,15 @@ public slots:
 
     QMap<OverlayID, Overlay::Pointer> _overlaysHUD;
     QMap<OverlayID, Overlay::Pointer> _overlaysWorld;
+#if OVERLAY_PANELS
     QMap<OverlayID, OverlayPanel::Pointer> _panels;
+#endif
     QList<Overlay::Pointer> _overlaysToDelete;
     unsigned int _stackOrder { 1 };
 
-    QReadWriteLock _lock;
```
Comment (jherico) [inline interface/src/ui/overlays/Overlays.h]: Locks no longer required since all operations take place on the main thread

**Comment by YuppyPlays** [comment]
Testing.

**Comment by YuppyPlays** [comment]
Testing complete:


Verify startup and shutdown function without hanging or crashing **Pass**
Verify Overlay functionality still works. (this script creates a number of overlays) **Pass**
image

Verify audio settings dialog still functions properly (check boxes work, changing input and output devices works) **Not sure if pass or fail, Enable noise reduction is the only thing not working but I am unsure if this works in current master?**
Smoke test functionality **Pass**
Monitor log output for warnings containing the string BlockingQueuedConnection (this doesn't fail the PR, it just informative and should give us clues to additional sources of race conditions) **Pass? I don't see any of these show up in the logs thus far.**

File: interface/src/Application.cpp [hunk 4]
```diff
@@ -1208,15 +1209,26 @@ Application::Application(int& argc, char** argv, QElapsedTimer& startupTimer, bo
...
+            // Stop the settings timer
+            settingsTimer->stop();
+            // Delete it (this will trigger the thread destruction
+            settingsTimer->deleteLater();
+            // Mark the settings thread as finished, so we know we can safely save in the main application 
+            // shutdown code 
+            _settingsGuard.trigger();
+        });
+
+        int SAVE_SETTINGS_INTERVAL = 10 * MSECS_PER_SECOND; // Let's save every seconds for now
```
Comment (sethalves) [inline interface/src/Application.cpp]: comment is wrong

File: interface/src/ui/overlays/Image3DOverlay.cpp [hunk 5]
```diff
@@ -45,11 +45,13 @@ Image3DOverlay::~Image3DOverlay() {
 }
 
 void Image3DOverlay::update(float deltatime) {
+#if OVERLAY_PANELS
```
Comment (sethalves) [inline interface/src/ui/overlays/Image3DOverlay.cpp]: Did you mean to disable this block of code?

**Pull request merged**